File: src/generators/derivation.ts
import { loadSpec } from "../core/spec";
import type { RootSpec } from "../core/types";

const ROOT_VOWELS = new Set(["A", "E", "I", "O", "U"]);

export interface DerivationResult {
  word: string;
  meaning: string;
  morphology: string;
  root: RootSpec;
  pattern: string;
  culturalNotes: string;
}

export function canonicalRoot(input: string): string {
  return input
    .toUpperCase()
    .replace(/[^A-Z]/g, "")
    .split("")
    .filter((char) => !ROOT_VOWELS.has(char))
    .join("");
}

export function findRoot(input: string): RootSpec | undefined {
  const wanted = canonicalRoot(input);
  return loadSpec().roots.roots.find((root) => {
    const ids = [root.id, root.form, ...(root.aliases ?? [])].map(canonicalRoot);
    return ids.includes(wanted);
  });
}

const patternAliases: Record<string, string> = {
  "root-verb": "verb",
  "verb": "verb",
  "noun": "noun",
  "root-noun": "noun",
  "adjective": "adjective",
  "agent": "agent",
  "object": "object",
  "ritual": "ritual",
  "ritual-poetic": "ritual",
  "civic": "civic",
  "civic-legal": "civic",
  "intimate": "intimate",
  "intimate-emotional": "intimate",
  "intimate-imperative": "intimate",
  "process": "process",
  "practice": "process",
  "ongoing": "process",
  "instrument": "instrument",
  "tool": "instrument",
  "place": "place",
  "field": "place",
  "doctrine": "doctrine",
  "theory": "doctrine",
  "collective": "collective",
  "guild": "collective",
  "lack": "lack",
  "absence": "lack",
  "negative": "lack",
  "category": "category",
  "class": "category",
  "discipline": "discipline",
  "training": "discipline",
  "office": "office",
  "mandate": "office",
  "record": "record",
  "archive": "record",
  "right": "right",
  "legal-right": "right",
  "vow": "vow",
  "oath": "vow"
};

export function deriveWord(rootInput: string, patternInput: string): DerivationResult {
  const root = findRoot(rootInput);
  if (!root) {
    throw new Error(`Unknown root '${rootInput}'`);
  }

  const pattern = patternAliases[patternInput] ?? patternInput;
  const derived = root.derived[pattern];

  if (!derived) {
    throw new Error(`Unknown pattern '${patternInput}'. Try: ${Object.keys(root.derived).join(", ")}`);
  }

  const imperativeNote =
    patternInput === "intimate-imperative"
      ? "Use with ke before the word for an explicit plea or command: ke " + derived.word + "."
      : "Ethra derivation keeps the semantic family visible inside the word.";

  return {
    word: derived.word,
    meaning: derived.meaning,
    morphology: `${root.id} root + ${derived.role} pattern`,
    root,
    pattern,
    culturalNotes: imperativeNote
  };
}

export function generateRoot(field: string): { root: string; rationale: string; sample: string } {
  const consonants = [
    ..."bcdfghjklmnprstvwxyz".split("")
  ];
  const cleaned = field.toLowerCase().replace(/[^a-z]/g, "");
  const picked: string[] = [];

  for (const char of cleaned) {
    if (consonants.includes(char) && !picked.includes(char)) {
      picked.push(char);
    }
    if (picked.length === 3) break;
  }

  while (picked.length < 2) {
    picked.push(consonants[(cleaned.length + picked.length * 5) % consonants.length]);
  }

  const root = picked.join("").toUpperCase();
  const sample = picked.map((c, index) => c + (index === picked.length - 1 ? "" : "a")).join("");
  return {
    root,
    sample,
    rationale: `Candidate root from prominent consonants in '${field}'. Add it to spec/roots.yaml before treating it as canonical.`
  };
}

File: src/derivation.rs
use crate::spec::load_spec;
use crate::types::RootSpec;
use anyhow::{Result, bail};
use serde::Serialize;
use std::collections::HashSet;

#[derive(Debug, Clone, Serialize)]
pub struct DerivationResult {
    pub word: String,
    pub meaning: String,
    pub morphology: String,
    pub root: RootSpec,
    pub pattern: String,
    #[serde(rename = "culturalNotes")]
    pub cultural_notes: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct GeneratedRoot {
    pub root: String,
    pub rationale: String,
    pub sample: String,
}

pub fn canonical_root(input: &str) -> String {
    let vowels = HashSet::from(['A', 'E', 'I', 'O', 'U']);
    input
        .to_uppercase()
        .chars()
        .filter(|char_value| char_value.is_ascii_uppercase())
        .filter(|char_value| !vowels.contains(char_value))
        .collect()
}

pub fn find_root(input: &str) -> Result<Option<RootSpec>> {
    let wanted = canonical_root(input);
    Ok(load_spec()?.roots.roots.iter().find_map(|root| {
        let mut ids = vec![root.id.clone(), root.form.clone()];
        ids.extend(root.aliases.clone());
        let matches = ids
            .iter()
            .map(|item| canonical_root(item))
            .any(|item| item == wanted);
        matches.then(|| root.clone())
    }))
}

fn pattern_alias(pattern_input: &str) -> &str {
    match pattern_input {
        "root-verb" | "verb" => "verb",
        "noun" | "root-noun" => "noun",
        "adjective" => "adjective",
        "agent" => "agent",
        "object" => "object",
        "ritual" | "ritual-poetic" => "ritual",
        "civic" | "civic-legal" => "civic",
        "intimate" | "intimate-emotional" | "intimate-imperative" => "intimate",
        "process" | "practice" | "ongoing" => "process",
        "instrument" | "tool" => "instrument",
        "place" | "field" => "place",
        "doctrine" | "theory" => "doctrine",
        "collective" | "guild" => "collective",
        "lack" | "absence" | "negative" => "lack",
        "category" | "class" => "category",
        "discipline" | "training" => "discipline",
        "office" | "mandate" => "office",
        "record" | "archive" => "record",
        "right" | "legal-right" => "right",
        "vow" | "oath" => "vow",
        _ => pattern_input,
    }
}

pub fn derive_word(root_input: &str, pattern_input: &str) -> Result<DerivationResult> {
    let Some(root) = find_root(root_input)? else {
        bail!("Unknown root '{}'", root_input);
    };
    let pattern = pattern_alias(pattern_input).to_string();
    let Some(derived) = root.derived.get(&pattern) else {
        let patterns = root.derived.keys().cloned().collect::<Vec<_>>().join(", ");
        bail!("Unknown pattern '{}'. Try: {}", pattern_input, patterns);
    };

    let cultural_notes = if pattern_input == "intimate-imperative" {
        format!(
            "Use with ke before the word for an explicit plea or command: ke {}.",
            derived.word
        )
    } else {
        "Ethra derivation keeps the semantic family visible inside the word.".to_string()
    };

    Ok(DerivationResult {
        word: derived.word.clone(),
        meaning: derived.meaning.clone(),
        morphology: format!("{} root + {} pattern", root.id, derived.role),
        root,
        pattern,
        cultural_notes,
    })
}

pub fn generate_root(field: &str) -> GeneratedRoot {
    let consonants = "bcdfghjklmnprstvwxyz".chars().collect::<Vec<_>>();
    let cleaned = field
        .to_lowercase()
        .chars()
        .filter(|char_value| char_value.is_ascii_lowercase())
        .collect::<String>();
    let mut picked: Vec<char> = Vec::new();

    for char_value in cleaned.chars() {
        if consonants.contains(&char_value) && !picked.contains(&char_value) {
            picked.push(char_value);
        }
        if picked.len() == 3 {
            break;
        }
    }

    while picked.len() < 2 {
        let index = (cleaned.len() + picked.len() * 5) % consonants.len();
        picked.push(consonants[index]);
    }

    let root = picked.iter().collect::<String>().to_uppercase();
    let sample = picked
        .iter()
        .enumerate()
        .map(|(index, char_value)| {
            if index == picked.len() - 1 {
                char_value.to_string()
            } else {
                format!("{}a", char_value)
            }
        })
        .collect::<String>();

    GeneratedRoot {
        root,
        sample,
        rationale: format!(
            "Candidate root from prominent consonants in '{}'. Add it to spec/roots.yaml before treating it as canonical.",
            field
        ),
    }
}
